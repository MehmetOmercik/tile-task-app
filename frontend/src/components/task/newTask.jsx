import { useDispatch, useSelector } from "react-redux";
import { updateEditTaskObject } from "./taskSlice";
import { DropdownComponent } from "../dropdownComponent";
import { putTask, postTask } from "../../utils/http";

export const NewTask = ({ setOverlaySection }) => {
  const dispatch = useDispatch();
  const { editTask: editTaskBoolean, editTaskObject } = useSelector(
    (state) => state.task
  );
  const { currentTileID } = useSelector((state) => state.tile);

  const handleProperty = (e, property) => {
    const value = e.target.value;

    dispatch(
      updateEditTaskObject({
        type: "UPDATE_PROPERTY",
        property,
        value,
      })
    );
  };

  const handleTypeOptionSelect = (selectedOption) => {
    console.log(selectedOption);
    dispatch(
      updateEditTaskObject({
        type: "UPDATE_PROPERTY",
        property: "type",
        value: selectedOption,
      })
    );
  };

  // const handleTileOptionSelect = (selectedOption) => {
  //   console.log(selectedOption);
  //   dispatch(
  //     updateEditTaskObject({
  //       type: "UPDATE_PROPERTY",
  //       property: "tile",
  //       value: selectedOption,
  //     })
  //   );
  // };

  const handleTaskSubmission = async (e) => {
    const taskPayload = {
      // id: editTaskObject.id
      title: editTaskObject.title,
      description: editTaskObject.description,
      order: +editTaskObject.order,
      type: editTaskObject.type,
      tile: currentTileID,
    };

    // e.preventDefault();
    editTaskBoolean
      ? putTask(editTaskObject.id, taskPayload)
      : postTask(taskPayload);

    // ! Doesnt update properly
    // window.location.reload();
    // await setOverlaySection("tasks");
  };

  return (
    <section className="bg-white  w-[300px] rounded-xl p-4">
      <h1 className="text-2xl mb-5">
        {editTaskBoolean
          ? `Update Task ${editTaskObject.id} for Tile ${currentTileID}`
          : `Create Task for Tile ${currentTileID}`}
      </h1>
      <form onSubmit={(e) => handleTaskSubmission(e)}>
        <label>Title</label>
        <input
          maxLength="30"
          value={editTaskObject.title}
          onChange={(e) => handleProperty(e, "title")}
          className="bg-gray-200 rounded-lg p-1 w-[97%]"
        ></input>

        <label htmlFor="">Description</label>
        <textarea
          value={editTaskObject.description}
          onChange={(e) => handleProperty(e, "description")}
          className="resize-none bg-gray-200 rounded-lg h-20 w-[97%] p-1"
          maxLength="256"
        ></textarea>

        <label htmlFor="">Order</label>
        <input
          value={editTaskObject.order}
          onChange={(e) => handleProperty(e, "order")}
          className="bg-gray-200 rounded-lg p-1 w-[97%] "
        ></input>

        <label>Type</label>
        <DropdownComponent
          options={["survey", "discussion", "diary"]}
          onSelect={handleTypeOptionSelect}
          defaultValue={editTaskObject.type}
        />
        {/* <label>Tile</label>
        <DropdownComponent
          options={[1, 2, 3]}
          onSelect={handleTileOptionSelect}
          defaultValue={editTaskObject.tile}
        /> */}
        <button
          type="submit"
          disabled={!editTaskObject.type}
          className="bg-slate-400 rounded-lg p-2 text-lg hover:bg-slate-300 disabled:bg-red-600"
        >
          {editTaskBoolean ? "Update" : "Create"}
        </button>
      </form>
    </section>
  );
};
