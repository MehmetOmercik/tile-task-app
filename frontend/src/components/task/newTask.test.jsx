import { NewTask } from "./newTask";
import { renderWithProviders } from "../../test/test-utils";

describe("New Task tests", () => {
  test("Snapshot test", () => {
    const newTask = renderWithProviders(<NewTask />);
    expect(newTask).toMatchSnapshot();
  });
});
