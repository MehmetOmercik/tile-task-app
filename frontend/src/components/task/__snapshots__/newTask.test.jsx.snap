// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`New Task tests > Snapshot test 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="bg-white  w-[300px] rounded-xl p-4"
      >
        <h1
          class="text-2xl mb-4"
        >
          Create Task for Tile 0
        </h1>
        <form>
          <label>
            Title
          </label>
          <input
            class="bg-gray-200 rounded-lg p-1 w-[97%] mb-3"
            maxlength="30"
            value=""
          />
          <label
            for=""
          >
            Description
          </label>
          <textarea
            class="resize-none bg-gray-200 rounded-lg h-20 w-[97%] p-1 mb-3"
            maxlength="256"
          />
          <label
            for=""
          >
            Order
          </label>
          <input
            class="bg-gray-200 rounded-lg p-1 w-[97%] mb-3"
            value="1"
          />
          <label>
            Type
          </label>
          <div
            class=""
          >
            <select
              class="p-1 rounded-md mb-4"
            >
              <option
                class=""
                value=""
              >
                Select an option
              </option>
              <option
                class=""
                value="survey"
              >
                survey
              </option>
              <option
                class=""
                value="discussion"
              >
                discussion
              </option>
              <option
                class=""
                value="diary"
              >
                diary
              </option>
            </select>
          </div>
          <button
            class="bg-slate-400 rounded-lg p-2 text-lg hover:bg-slate-300 disabled:bg-red-600 mb-1"
            disabled=""
            type="submit"
          >
            Create
          </button>
        </form>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="bg-white  w-[300px] rounded-xl p-4"
    >
      <h1
        class="text-2xl mb-4"
      >
        Create Task for Tile 0
      </h1>
      <form>
        <label>
          Title
        </label>
        <input
          class="bg-gray-200 rounded-lg p-1 w-[97%] mb-3"
          maxlength="30"
          value=""
        />
        <label
          for=""
        >
          Description
        </label>
        <textarea
          class="resize-none bg-gray-200 rounded-lg h-20 w-[97%] p-1 mb-3"
          maxlength="256"
        />
        <label
          for=""
        >
          Order
        </label>
        <input
          class="bg-gray-200 rounded-lg p-1 w-[97%] mb-3"
          value="1"
        />
        <label>
          Type
        </label>
        <div
          class=""
        >
          <select
            class="p-1 rounded-md mb-4"
          >
            <option
              class=""
              value=""
            >
              Select an option
            </option>
            <option
              class=""
              value="survey"
            >
              survey
            </option>
            <option
              class=""
              value="discussion"
            >
              discussion
            </option>
            <option
              class=""
              value="diary"
            >
              diary
            </option>
          </select>
        </div>
        <button
          class="bg-slate-400 rounded-lg p-2 text-lg hover:bg-slate-300 disabled:bg-red-600 mb-1"
          disabled=""
          type="submit"
        >
          Create
        </button>
      </form>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
